### DISCUSSION TITLE
there is an official python LSP library from microsoft

### DISCUSSION THREAD

**Discussion by llllvvuu** [Ackee-Blockchain/wake#270] [Technical]
[microsoft/lsprotocol (`lsprotocol` on pip)](https://github.com/microsoft/lsprotocol) can be used in order to not have to maintain type definitions like [these](https://github.com/Ackee-Blockchain/woke/blob/main/woke/lsp/common_structures.py).

Also, there is [pygls](https://github.com/openlawlibrary/pygls) which is [decently adopted](https://github.com/openlawlibrary/pygls/blob/main/Implementations.md) and takes care of other boilerplate such as transport ([tcp/stdio/ws](https://pygls.readthedocs.io/en/latest/pages/advanced_usage.html#connections)), [caching, document state, and lifecycle events](https://pygls.readthedocs.io/en/latest/pages/advanced_usage.html#built-in-features).

If it ain't broke don't fix it but I thought I'd offer the tip anyway in case it comes in handy later ([I was able to make a - granted much simpler - LSP in <120 LOC with this](https://github.com/llllvvuu/Fixit/blob/feat/lsp/src/fixit/lsp.py)).

**Comment by hacker-DOM** [comment]
Nice!
